src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine_FindReferencesInDocuments.cs

```diff
@@ -101,16 +101,21 @@ async ValueTask PerformSearchInProjectAsync(ImmutableArray<ISymbol> symbols, Pro
         async ValueTask PerformSearchInDocumentWorkerAsync(
             ISymbol symbol, FindReferencesDocumentState state)
         {
+            // Scratch buffer to place references for each finder. Cleared at the end of every loop iteration.
+            using var _ = ArrayBuilder<FinderLocation>.GetInstance(out var referencesForFinder);
+
             // Always perform a normal search, looking for direct references to exactly that symbol.
             foreach (var finder in _finders)
             {
-                var references = await finder.FindReferencesInDocumentAsync(
-                    symbol, state, _options, cancellationToken).ConfigureAwait(false);
-                foreach (var (_, location) in references)
+                await finder.FindReferencesInDocumentAsync(
+                    symbol, state, StandardCallbacks<FinderLocation>.AddToArrayBuilder, referencesForFinder, _options, cancellationToken).ConfigureAwait(false);
+                foreach (var (_, location) in referencesForFinder)
                 {
                     var group = await ReportGroupAsync(symbol, cancellationToken).ConfigureAwait(false);
                     await _progress.OnReferenceFoundAsync(group, symbol, location, cancellationToken).ConfigureAwait(false);
                 }
+
+                referencesForFinder.Clear();
```

Comment:
it's interesting how close this pattern is to the pipewriting/reading stuff i just did.  this might get cleaner (in the future) with a channel that is pulling off the items and reporting to them to the progress, while the main find is pushing into the channel.